### PR TITLE
Fixed #24424 - sqlite schema editor causes SQL syntax error

### PR DESCRIPTION
**Pull request by adnam**
Resolved problem described in [Ticket 24424](https://code.djangoproject.com/ticket/24424) and provided unit test.

### CODE REVIEW

**Comment by MarkusH** [comment]
Hey @adnam . Thanks for the patch. I don't think the test is quite correct, though. You're creating a MTI model layout and then removing the `_ptr` from the inheriting model, making that model completely invalid.

If a model inherits from another concrete model it will have a `_ptr` field. If it inherits from an abstract model or `models.Model` it has its `id = AutoField()` field (unless the abstract model overrides that). Thus, altering the bases implicitly adds a `id` column and removes `_ptr`.

Can you add an example project to the ticket that shows the problem you are faced with.

**Comment by adnam** [comment]
Hi @MarkusH,

I discovered this in a proprietary project and it took me a while to figure out what was happening. Essentially there was a model extending an abstract model, which was subsequently modified to extend a different model. The migration was generated with `manage.py makemigrations` which generated a series of operations. I managed to boil it down to [3 steps](https://github.com/adnam/django/commit/bc29277fabb3f7fa5097ce21ccaea32c271bb8b8#diff-5210414c7a29d5d1f61588388b517599R1862) which gave an identical SQL syntax error. I agree it's a bit of a weird one.

**Comment by MarkusH** [comment]
Thanks for the explanation, but that sounds more like a bug in the autodetector constructing operations that, when applied, generate invalid models.

**Comment by adnam** [comment]
I honestly don't know, but what I saw is that the migration worked using a postgres database but failed with sqlite.

I think this is due to sqlite's limited `ALTER TABLE` support compared to other DMBs, so the schema editor re-creates the whole table as a midway step. The SQL that failed was copying over the values from the old table to the new one: `INSERT INTO "ramptr_unicorn__new" () SELECT FROM "ramptr_unicorn";`

**Comment by adnam** [comment]
Hi @MarkusH, just to clarify further:

I think it's not quite the case that the autodetector is creating operations that create invalid models. Rather, the migration has some intermediary steps for which the sqlite schema editor may generate invalid SQL. Subsequent operations put the model back to a "normal" state, but migrations are applied atomically this is rarely seen.

**Comment by MarkusH** [comment]
I thought about this problem a bit more: since it's indeed the SQLite3 schema editor that poses a problem, I rather see an explicit test in `tests/schema/tests.py`, e.g. `test_remove_last_field()`:

```python
def test_remove_last_field(self):
    class TestModel(models.Model):
        field = models.IntegerField(primary_key=True)

        class Meta:
                app_label = 'schema'
                apps = new_apps

    self.local_models = [TestModel]

    # Create model with schema editor
    # Verify columns: connection.cursor().introspection.get_table_description() (see get_indexes())
    # Remove field with schema editor
    # Verify columns: connection.cursor().introspection.get_table_description() (see get_indexes())
```

Given this test case, I don't think we would even need to limit it to SQLite. It should work fine on all other backends as well.

**Comment by MarkusH** [comment]
Can you also add a release note to `docs/releases/1.7.6.txt` since I intend to backport the patch.

File: docs/releases/1.7.6.txt [hunk 1]
```diff
@@ -9,4 +9,4 @@ Django 1.7.6 fixes several bugs in 1.7.5.
 Bugfixes
 ========
 
-* ...
+* SQLite schema editor causes SQL syntax error (:ticket:`24424`).
```
Comment (MarkusH) [inline docs/releases/1.7.6.txt]: Can you use past tense please. Something like ```Fixed a migration crash when removing the last database table column on SQLite (:ticket:`24424`).```

**Comment by MarkusH** [comment]
@adnam Could you replace your unittest with one as described in https://github.com/django/django/pull/4223#issuecomment-77330629

**Comment by MarkusH** [comment]
buildbot, test this please.

File: tests/schema/tests.py [hunk 2]
```diff
@@ -1371,3 +1371,34 @@ def test_alter_field_default_dropped(self):
...
+        self.local_models = [TestModel]
+
+        # Create model with schema editor
+        with connection.schema_editor() as editor:
+            editor.create_model(TestModel)
+
+        test_field = TestModel._meta.get_field("test_field")
+
+        # Ensure the table is there and has the right index
+        self.assertIn(
```
Comment (timgraham) [inline tests/schema/tests.py]: these assertions would fit on 1 one

**Comment by timgraham** [comment]
Could you squash commits and move release notes to 1.7.7? Test failure on MySQL also needs to be fixed or skipped.

**Comment by timgraham** [comment]
Could you try to [squash your commits](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/working-with-git/#rebasing-branches)?

**Comment by adnam** [comment]
@timgraham, welp to sqash all the commits into one I had to delete a re-create [the ticket_24424 branch](https://github.com/adnam/django/commits/ticket_24424) and now my pull-request has been closed. Should I open a new one?

**Comment by timgraham** [comment]
Yes, it seems like you have to in that case. In the future, you can use force push to update the branch, rather than deleting it.

**Comment by adnam** [comment]
OK, I give up.

**Comment by adnam** [comment]
@timgraham, I followed the instructions on git-scm.com, and did not know that it was going to mess up the pull-request. If you want the bugfix help yourselves: https://github.com/adnam/django/tree/ticket_24424

Cheers, Adam

**Comment by timgraham** [comment]
You need to rebase on master as some notes have been added to 1.7.7 since you created your branch. Did you resolve the issue that Markus raised: https://code.djangoproject.com/ticket/24424#comment:11?

**Comment by adnam** [comment]
Hi @timgraham,

@MarkusH made the comment that "the fundamental problem is ... the autodetector generates models without any fields", which is possibly true, but I'm afraid I'm unable to fix that without a good deal of time (which I don't have).

The specific problem I found was with the sqlite schema editor': with limited `alter table` support it has to re-create tables and copy over the data. It was this copy-sql query that gave a problem in a unique edge case - I discovered this when a MTI child model was changed to extend a different parent.

I completely forked up this pull request so I created a new on (#4325), hopefully it's OK now!

**Comment by robkane1** [comment]
Hi all - realise this is very old, but stumbled across this looking for a solution to this problem which still shows itself in Django 1.8.9 with SQLite when deleting a model which subclasses another model and has a ManyToMany with Sites. So for the benefit of anybody else who does the same and has a sore head from banging it....

My guess is that this is something to do with the ordering of the save process, but to be honest this is beyond my scope.

**Deleting the 'remove field' instructions from the migration file, leaving just the 'delete model' instruction seems to resolve this and allows the migration to run without issue.** I don't know if this will have any knock on effects, but given that the model itself has been deleted (or moved) it follows that the fields have done the same. 

Not sure how the state of migrations is during the process, but is a remove field even necessary when a delete model is present in the same migration? You wouldn't do this in raw SQL, but perhaps I am missing something important.  

Here is the traceback from my issue:

```
Traceback (most recent call last):
  File "manage.py", line 10, in <module>
    execute_from_command_line(sys.argv)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
    utility.execute()
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/core/management/__init__.py", line 346, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/core/management/base.py", line 394, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/core/management/base.py", line 445, in execute
    output = self.handle(*args, **options)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/core/management/commands/migrate.py", line 222, in handle
    executor.migrate(targets, plan, fake=fake, fake_initial=fake_initial)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/migrations/executor.py", line 110, in migrate
    self.apply_migration(states[migration], migration, fake=fake, fake_initial=fake_initial)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/migrations/executor.py", line 148, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/migrations/migration.py", line 115, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/migrations/operations/fields.py", line 121, in database_forwards
    schema_editor.remove_field(from_model, from_model._meta.get_field(self.name))
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/backends/sqlite3/schema.py", line 197, in remove_field
    self._remake_table(model, delete_fields=[field])
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/backends/sqlite3/schema.py", line 147, in _remake_table
    self.quote_name(model._meta.db_table),
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/backends/base/schema.py", line 111, in execute
    cursor.execute(sql, params)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/backends/utils.py", line 79, in execute
    return super(CursorDebugWrapper, self).execute(sql, params)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/utils.py", line 98, in __exit__
    six.reraise(dj_exc_type, dj_exc_value, traceback)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/utils/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/rob/Python/puredj/venv/lib/python3.4/site-packages/django/db/backends/utils.py", line 64, in execute
    return self.cursor.execute(sql, params)
```

**Pull request closed without merging**